Refactor background fills and image copy

pkohut opened the pull request:
(Replaces PR #108)

Cleans up code that does the background fill operations and functions which copy pngwriter image data
from source object to destination object (graph_).  Also, updated the operator = to a near proper
copy-and-swap idiom (note added in comments on making it fully proper and C++11 ready).

Here's the before and after benchmarks of running "performance.cc" in release mode with an
image size of 10000 x 10000 and background color set to 4.

```
                        Before      After
                       Refactor    Refactor   Improvement
-----------------------+----------+----------+-----------
allocation time:       |   928461 |   472277 | 0.5087
random value time:     |  2352986 |  2275185 | 0.9669
png copy (in RAM) time:|   915583 |   689525 | 0.7531
png scale_k(0.5) time: |  2550104 |  2500438 | 0.9805
png scale_k(2.0) time: | 40625981 | 39492986 | 0.9721
write_png() time:      | 25782479 | 26071078 | 1.0112
```

Will continue to provide updated benchmarks as refactoring continues.  

I don't think the changes made break ABI compatibility.  If that's a concern, maybe someone can test for that.  Once I get more into the project and get some compiled samples put together I can check for breakage then.

ax3l commented on src/pngwriter.h:
```diff
@@ -196,6 +200,8 @@ class pngwriter
     * */
    pngwriter & operator = (const pngwriter & rhs);
 
+   friend void swap(pngwriter & lhs, pngwriter & rhs);
```
can you pls add (doxygen) header comments?

Why don't you implements this as a private member?

pkohut replied:
Added comments.

As for private/public...Your call.  Made them private.

ax3l commented on src/pngwriter.h:
```diff
@@ -732,6 +738,12 @@ class pngwriter
 
    int static get_text_width_utf8(char * face_path, int fontsize, char * text);
 
+   int fillBackgroundColor(void);
```
Can you please add (doxygen) header comments for all of those?

Should they really be public (aka, why do we need to extent the API)?

pkohut committed:
Refactor background fills and image copy

Fix mapping of RGB565 and RGB555 to RGB888
http://stackoverflow.com/questions/2442576/how-does-one-convert-16-bit-rgb565-to-24-bit-rgb888/9069480#9069480

1) Background fills are now done with fillBackgroundColor(void) which then
calls fillBackgroundWithColor(int).

This is more performant, as it removes all dereferencing of
backgroundcolour_ from the heavy loop.

Origianl code
./performance   10000 x 10000 image with background color = 4
allocation time: 928461
random value time: 2352986
png copy (in RAM) time: 915583
png scale_k(0.5) time: 2550104
png scale_k(2.0) time: 40625981
write_png() time: 25782479

After this refactor
./performance   10000 x 10000 image with background color = 4
allocation time: 473646
random value time: 2342991
png copy (in RAM) time: 960807
png scale_k(0.5) time: 2557795
png scale_k(2.0) time: 41130223
write_png() time: 26498281

2) Refactor duplicate code that copies data from source image to destination
graph_.

Again, slight perf improvement over last commit.
./performance     10000 x 10000 image background color = 4
allocation time: 478769
random value time: 2307549
png copy (in RAM) time: 685744
png scale_k(0.5) time: 2508808
png scale_k(2.0) time: 39133261
write_png() time: 26023528

3) Refactor operator = to copy-and-swap. Neglable performance difference
from prior commit.

./performance     10000 x 10000 image background color = 4
allocation time: 472277
random value time: 2275185
png copy (in RAM) time: 689525
png scale_k(0.5) time: 2500438
png scale_k(2.0) time: 39492986
write_png() time: 26071078

------

Tighten up copyImageDataFrom code for a 47% speed boost.

Use png_color_struct for better encapsulation of color values.

Use memset for background fill when all channels are equal.

Add function to decode RGB channel values from 16 bit color input.

Make new member function private.

pkohut commented:
Update PR to fix mapping of RGB565 and RGB555 to RGB888

http://stackoverflow.com/questions/2442576/how-does-one-convert-16-bit-rgb565-to-24-bit-rgb888/9069480#9069480
